app/models/concerns/facets/managed_host_extensions.rb

```diff
@@ -0,0 +1,66 @@
+require 'facets'
+
+module Facets
+  module ManagedHostExtensions
+    extend ActiveSupport::Concern
+
+    included do
+      Facets::ManagedHostExtensions.refresh_facet_relations(self)
+
+      def attributes
+        hash = super
+
+        # include all facet attributes by default
+        facets_with_definitions.each do |facet, facet_definition|
+          hash["#{facet_definition.name}_attributes"] = facet.attributes.reject { |key, _| %w(created_at updated_at).include? key }
+        end
+        hash
+      end
+    end
+
+    def self.refresh_facet_relations(klass)
+      Facets.registered_facets.values.each do |facet_config|
+        self.register_facet_relation(klass, facet_config)
+      end
+    end
+
+    def self.register_facet_relation(klass, facet_config)
+      klass.class_eval do
+        has_one facet_config.name, :class_name => facet_config.model.name, :foreign_key => :host_id, :inverse_of => :host
+        accepts_nested_attributes_for facet_config.name
+        alias_method "#{facet_config.name}_attributes", facet_config.name
+
+        include facet_config.extension if facet_config.extension
+
+        include_in_clone facet_config.name
+
+        facet_config.compatibility_properties.each do |prop|
+          define_method(prop) { |*args| forward_property_call(prop, args, facet_config.name) }
+        end if facet_config.compatibility_properties
+      end
+    end
+
+    def facets
+      facets_with_definitions.keys
+    end
+
+    # This method will return a hash of facets for a specific host including the coresponding definitions.
+    # The output should look like this:
+    # { host.puppet_aspect => Facets.registered_facets[:puppet_aspect] }
+    def facets_with_definitions
```

Comment:
Nitpick, if what you're returning is the `facet_config`, why calling it `facets_with_definitions`? Maybe calling this just `facets`, and renaming `facet_names` above would work better. 

In any case when one calls `.something_in_plural` in Rails we normally expect it to be an Array of relations (i.e: `host.environments` would return an array of environments, not a hash of `environment.name => environment`). How about then returning an array of facets?